### DISCUSSION TITLE
Lazy initialized non-nullable auto properties

### DISCUSSION THREAD

**Discussion by popcatalin81** [dotnet/csharplang#7145] [Language Ideas]
Currently non-nullable auto properties present a eager initialization problem when they are supposed to have default value. This can lead to redundant initialization and unnecessary GC pressure. 

For example a mutable class representing a hypothetical product, specifically designed for non-nullable properties:
```csharp
public class Product
{
    public required string Name { get; set; }
    public List<Category> Categories { get; set; } = new() { Category.Default };
    public List<string> Tags { get; set; } = new();
}
```
An a typical initialization:
```csharp
var p = new Product() 
{
    Name = "Peaches",
    Categories = new() { foods, fruits },
    Tags = new () { "Summer", "Sweet", "Fresh" },
};
```

The default values for the `Categories` and the `Tags` properties are redundant in this scenario and only create garbage, Two Lists are created and immediately discarded, becoming GC eligible. 

Currently the way to avoid this redundant initialization is to convert auto properties to full properties and to use lazy initialization:

```csharp
public class Product
{
    private List<Category>? categories;
    private List<string>? tags;

    public required string Name { get; set; }
    public List<Category> Categories 
    { 
        get => categories ??= new() { Category.Default }; 
        set => categories = value; 
    }

    public List<string> Tags 
    { 
        get => tags ??= new(); 
        set => tags = value; 
    }
}
```
The proposal is to allow the `??=` null coalescing assignment operator directly on auto properties initializer like this:

```csharp
public class Product
{
    public required string Name { get; set; }
    public List<Category> Categories { get; set; } ??= new() { Category.Default };
    public List<string> Tags { get; set; } ??= new();
}
```

This will make lazy initialization for non-nullable auto properties quite easy to implement.

**Comment by huoyaoyuan** [comment]
This can actually be covered by post-initialization logic that runs after initializer block. Currently any post-logic can only be achieved with constructor and its parameters.

**Reply by DavidArno** [reply]
> This can actually be covered by post-initialization logic that runs after initializer block. 

That wouldn't cover this use-case properly. It would still result in eager initialization. That initialization would be delayed simply until that post-initialization logic ran, but that would still be part of the object initialization.

The aim here is to achieve lazy initialization (only initialized when called) of properties without all the boilerplate ceremony that is currently required.

**Comment by HaloFour** [comment]
With [semi-auto properties](https://github.com/dotnet/csharplang/issues/140) this could be achievable with this:

```cs
    public List<Category> Categories { set; get => field ??= new() { Category.Default }; }
    public List<string> Tags { set; get => field ??= new(); }
```

**Reply by popcatalin81** [reply]
While that's true and ignoring the other issues semi-auto properties have, I would prefer not ho have a separate mechanism for lazy initialization vs eager initialization of auto properties:

A simple switch from: 
```csharp
public List<string> Tags { get; set; } = new();
```
to this:
```csharp
public List<string> Tags { get; set; } ??= new();
```
Would change the initialization from eager to lazy, without refactoring from auto properties to semi-auto properties.

**Reply by HaloFour** [reply]
> I would prefer not ho have a separate mechanism for lazy initialization

Syntax aside, the problem with all of the lazy initialization proposals is that they can't address the policy questions around thread safety.

**Reply by popcatalin81** [reply]
The manual lazy initialized version:
```csharp
    public List<string> Tags 
    { 
        get => tags ??= new(); 
        set => tags = value; 
    }
```
Is not thread safe either, yet necessary pattern to avoid unnecessary object creation and GC pressure. Mutable structures never offered thread safety guarantees, I would not expect this implementation to even try.

**Reply by HaloFour** [reply]
> I would not expect this implementation to even try.

I would expect a language feature that encourages this behavior to not be wrong two-thirds of the time, which it will be if any specific policy opinion is baked into the language.

**Reply by colejohnson66** [reply]
For thread safety, C#, unlike the new kid named Rust, makes thread-safety opt-in. The language provides almost no support for thread safety outside `lock` being lowered to `Monitor.Enter/Exit`; All thread-safe features are in the BCL. As a result, I'd expect that a `??=` syntax as proposed would, by default, be thread-unsafe, and, if one wanted thread-safety, they'd use some other data structure than a raw property.

**Reply by popcatalin81** [reply]
Mutable property access is not assumed to be thread safe anywhere in the frameworlk regardless if the property is eager or lazy initialized :

This is not a thread safe property:
```csharp
  public int? Count { get;set } = 1;
```
and neither would this be:
```csharp
  public int? Count { get;set } ??= 1;
```
In both cases accessing the value without synchronization is not safe.

**Reply by HaloFour** [reply]
> s a result, I'd expect that a ??= syntax as proposed would, by default, be thread-unsafe, and, if one wanted thread-safety, they'd use some other data structure than a raw property.

I wholeheartedly disagree.  The fact that the language doesn't offer policy control for a language feature that does not exist does not mean that the language should ignore the inherent issues of policy in a proposal that seeks to add that feature.  Lazy initialization is currently something achieved through BCL classes, which do explicitly offer three different thread safety policies.

> Mutable property access is not assumed to be thread safe anywhere in the frameworlk regardless if the property is eager or lazy initialized

Mutation of the property isn't thread-safe, unless the property implements thread-safety within the accessor.  But we're not talking about setter accessor methods, we're talking about initialization, which occurs during construction before the base constructor is called, and is effectively thread safe, unless you go out of your way to write pathological code.

**Reply by HaloFour** [reply]
And, IMO, that's another point against the proposed syntax.  Initializer expressions are expected to run exactly one time during the construction of the class.  This is not an initializer, thus should not use initializer syntax.  It should use the existing syntax for declaring an expression that runs when the property is accessed, which is the following:

```cs
public int Count { get => 1; set; }
```

Which, when combined with a null conditional assignment, already looks like this:

```cs
private int? _count;
public int Count { get => _count ??= 1; set => _count = value; }
```

Which, when modified for support for semi-auto properties, looks like this:

```cs
public int Count { get => field ??= 1; set; }
```

This is even one of the motivating use cases for semi-auto properties, and it enables the developer to express the desired thread safety policies within the expression of the getter accessor method.

**Reply by HaloFour** [reply]
> For thread safety, C#, unlike the new kid named Rust, makes thread-safety opt-in.

There isn't a single way to opt-in to thread safety.  Each strategy trades performance for stricter control over execution and publication.  This matters if it's undesirable for the initializer to run more than once, or for it to be possible for two consumers to end up with two separate results.  If that property were to return a mutable `List<T>` and two separate consumers mutated it they could be affecting change to completely separate lists.

**Reply by CyrusNajmabadi** [reply]
I would expect lazy initialization to be thread safe, like it is when accessing statics on a type.

**Reply by popcatalin81** [reply]
I don't want thread safety for lazy initialization because the simple full property scenario:
```csharp
    public List<Category> Categories 
    { 
        get => categories ??= new() { Category.Default }; 
        set => categories = value; 
    }
```
doesn't offer it either. That's not thread safe, and not guaranteed for the lazy initialization to run once. 
If I want guaranteed single time initialization then, I will opt in to it, by using synchronization, (Lazy<T> is not an option as it can't be used for mutable properties) :

```charp
    public List<Category> Categories
    {
        get
        {
            if (categories == null)
                lock (lockObject)
                    if (categories == null)
                        categories = new() { Category.Default };
            return categories;
        }

        set => categories = value;
    }
```
This offers guaranteed zero or one time initialization, but the property is still not thread safe either way. 

Also with the shorthand:
```csharp
public List<Category> Categories { get; set; } ??= new() { Category.Default };
```
No new thread safety problems are being introduced in the language. The thread safety problems, like multiple initialization already exist in the lazy pattern itself, which is currently in use:
```csharp
    public List<Category> Categories 
    { 
        get => categories ??= new() { Category.Default }; 
        set => categories = value; 
    }
```
I don't need the shorthand syntax to solve the lazy initialization problems of the pattern it's a shorthand for. I just want a simple way to do lazy initialization with defaults of auto properties.

**Reply by HaloFour** [reply]
> I don't want thread safety for lazy initialization because the simple full property scenario

That is something you've chosen by virtue of writing your accessor method to not be thread safe.  That is _very_ different from the language making that decision for everyone.

> I don't need the shorthand syntax to solve the lazy initialization problems of the pattern it's a shorthand for. I just want a simple way to do lazy initialization with defaults of auto properties.

Everyone has different needs.  In many languages that have features similar to this (like with static locals), it's expected that the initialization is threadsafe.

**Reply by popcatalin81** [reply]
> I would expect lazy initialization to be thread safe, like it is when accessing statics on a type.

Lazy initialized initialized statics are not thread safe either unless synchronization is used or thread safe Lazy<T>. 

```csharp
public static Category Default => defaultCategory ??= new Category();
```

Yes, this is not a pattern recommended by the framework design guidelines, which recommends thread safety for static members, except for mutable static fields, which are not thread safe anyway (because they are mutable)

Also there's less reason to do lazy initialization for static fields than it is for instance fields. There's no memory to be saved, as once initialized the lifetime of the instance  is till the end of the process/app domain. The only benefit would be startup, and if it's an costly initialization routine, then you would want to use some sort of synchronization mechanism to ensure the lazy synchronization runs only once. And besides this static field initialization using an initializer is already in a sense lazy. As the runtime can delay it's initialization until it's accessed.

**Reply by HaloFour** [reply]
> Lazy initialized initialized statics are not thread safe either

Again, you've chosen to make the accessor not thread safe.  But that's not what Cyrus was referring to.  He was talking about accessing static members, which are lazy initialized by the runtime when the type is loaded, and the runtime _does_ guarantee that this is threadsafe.

**Reply by popcatalin81** [reply]
> Everyone has different needs. In many languages that have features similar to this (like with static locals), it's expected that the initialization is threadsafe.

C# already has plenty of features that it chose not implement as thread safe. Do you have specific scenarios where you'd need lazy initialization to be thread safe on a mutable property in mind? Why would you need  lazy initialization to be thread safe for a mutable property that is not supposed to be thread safe anyway ??

**Reply by CyrusNajmabadi** [reply]
> Why would you need lazy initialization to be thread safe

To match the existing lazily initialized members already in the language which *are* threadsafe. It would be strange to me to add this and not behave the same way.

**Reply by popcatalin81** [reply]
null coalescing assignment operator `??=` is not expected to be thread safe.

> > Why would you need lazy initialization to be thread safe
> 
> To match the existing lazily initialized members already in the language which _are_ thread safe. It would be strange to me to add this and not behave the same way.

I don't think there are any noteworthy examples of read write properties in the language that are thread safe, respective of how they are initialized.

Read only properties on the other hand are expected to be thread safe (unless specified otherwise);

This is a very common pattern for lazy initialization for properties which are specified as not being thread safe (Which is the default guideline for public instance properties)

Before null coalescing operator this was used:
```csharp
    public List<Category> Categories
    {
        get
        {
            if (categories == null)
                categories = new List<Category>();
            return categories;
        }
        set => categories = value;
    }
```

Before null coalescing assignment operator this was used
```csharp
    public List<Category> Categories
    {
        get
        {
           return categories ?? (categories = new List<Category>());
        }

        set => categories = value;
    }
```

With null coalescing assignment operator this is used now:
```csharp
    public List<Category> Categories
    {
        get => categories ??= new List<Category>();
        set => categories = value;
    }
```

These are extremely common patterns in .Net everywhere. And since the properties are not expected to be thread safe, most of the time there's no point in making the initialization thread safe either.

**Reply by HaloFour** [reply]
> Read only properties on the other hand are expected to be thread safe

I would definitely argue that initialization behavior should not differ based on whether or not a property had a set/init accessor.  In all of those cases you've demonstrated, the developer explicitly opted into having a non-threadsafe behavior.  If that's what the developer wants, that's fine, but that shouldn't be the behavior espoused and encouraged by a specific language feature.

The proposed implementation via null conditional assignment also assumes that `null` is not a valid initialized value, which will continually trigger re-initialization.  That might be fine for your specific use case, but that's not fine in general.  It's expected that the initialization should happen exactly once, and can't be triggered again through an errant reassignment.

**Reply by popcatalin81** [reply]
I don't think `Initialization` and `Lazy Initialization` should necessarily have the same semantics.I think that's an unreasonable expectation. The `LazyThreadSafetyMode` enumeration of `Lazy<T>`  has 3 values:

- ExecutionAndPublication
- None
- PublicationOnly

Which shows that while eager initialization is always thread safe, lazy initialization may not be, and may not guarantee single execution either. That's a choice of the implementer. 

> The proposed implementation via null conditional assignment also assumes that null is not a valid initialized value, which will continually trigger re-initialization

Null is not a valid value since the property is non-nullable. 

> It's expected that the initialization should happen exactly once

True for eager initialization. Not true for lazy initialization unless implemented that way, either using thread safe Lazy<T> with LazyThreadSafetyMode.ExecutionAndPublication or other pattern. If this were true, Lazy<T> would not offer any other options except the thread safe single execution one.

**Reply by HaloFour** [reply]
> I don't think `Initialization` and `Lazy Initialization` should necessarily have the same semantics.I think that's an unreasonable expectation.

It would be surprising and unintuitive for lazy initialization to be less safe.  It will lead to subtle and unexpected bugs.

> Which shows that while eager initialization is always thread safe, lazy initialization may not be, and may not guarantee single execution either. That's a choice of the implementer.

Right, that is a choice of the _implementer_.  The language should never make an assumption that any of those three policies are the appropriate choice for any given use case, nor should the language encourage one over any other via special initialization syntax.

> Null is not a valid value since the property is non-nullable.

NRTs don't affect runtime behavior.

> True for eager initialization. Not true for lazy initialization unless implemented that way, either using thread safe Lazy with LazyThreadSafetyMode.ExecutionAndPublication or other pattern. If this were true, Lazy would not offer any other options except the thread safe single execution one.

Initialization should happen once regardless of when it happens.  That's why it's initialization and not simply null coalescing.  That's why `LazyInitializer` takes an `ref bool initialized` flag.  Assumptions around sentinel values should definitely not be baked into a language feature.

**Reply by popcatalin81** [reply]
> Initialization should happen once regardless of when it happens.

Again you're talking about eager initialization. 

In fact this never happens in lock free data structures. Lock free data structures DO NOT guarantee single execution of lazy initializers, not can they (Lock free meaning, no monitor and no spin locks either). In fact lock free data structures work on the principle, first thread to finish the initialization wins and it can store the results, the other will discard them.

**Reply by HaloFour** [reply]
> Again you're talking about eager initialization.

Initialization is initialization.

> Lock free data structures DO NOT guarantee single execution of initializers, not can they (Lock free meaning, no monitor and no spin locks either). In fact lock free data structures work on the principle, first thread to finish the initialization wins and it can store the results, the other will discard them.

That they do, which is why it's one of the options when using `Lazy<T>`, `LazyThreadSafetyMode .PublicationOnly`.  But CAS strategies does guarantee that the property could only ever return the same instance.  That's only one strategy, though, and it's certainly not the only appropriate strategy.

**Reply by popcatalin81** [reply]
Sorry I meant execution above.

> That they do, which is why it's one of the options when using Lazy<T>, LazyThreadSafetyMode .PublicationOnly. But CAS strategies does guarantee that the property could only ever return the same instance.

`LazyThreadSafetyMode .PublicationOnly` does not guarantee single execution, only single initialization.

Then what if 
```csharp
public List<Category> Categories { get; set; } ??= new() { Category.Default },;
```
is equivalent to:

```csharp
    public List<Category> Categories
    {
        get
        {
            if (categories == null)
                Interlocked.CompareExchange(ref categories, new() { Category.Default }, null);
            return categories;
        }

        set => categories = value ?? throw new NullReferenceException();
    }
```
This satisfies your single initialization requirement. Even though I fail to see how this is better than the simpler option.

**Reply by HaloFour** [reply]
> `LazyThreadSafetyMode .PublicationOnly` does not guarantee single execution, only single initialization.

I know, it's a CAS approach where the first execution wins.  If you want to guarantee single execution you would use `LazyThreadSafetyMode.ExecutionAndPublication`.

**Reply by jnm2** [reply]
It seems reasonable to me to say that the syntax `??=` implies the very specific lazy init strategy of `x ?? (x =`, and thus is explicitly non-threadsafe. I would have an icky feeling if the syntax `??=` came with something heavier than what `??=` means in every executable context today.

**Reply by jnm2** [reply]
> While that's true and ignoring the other issues semi-auto properties have, I would prefer not ho have a separate mechanism for lazy initialization vs eager initialization of auto properties:
> 
> A simple switch from:
> 
> ```cs
> public List<string> Tags { get; set; } = new();
> ```
> 
> to this:
> 
> ```cs
> public List<string> Tags { get; set; } ??= new();
> ```
> 
> Would change the initialization from eager to lazy, without refactoring from auto properties to semi-auto properties.

@popcatalin81 I understand what you're asking. If enough people felt that *reading* the `field`-based syntax was too onerous, then a new special shorthand syntax could be considered. The _writing_ (refactoring) however is something you can have almost for free when #140 ships (or today if you generate a field) by writing a Roslyn code refactoring and installing it as a VSIX or NuGet package. A quick light bulb invocation could swap between the lazy and eager forms.

```cs
public List<string> Tags { get; set; } = new();

// 💡 > Lazily initialize:
public List<string> Tags { get => field ??= new(); set; }

// Big enough savings for a common enough case in enough domains?
public List<string> Tags { get; set; } ??= new();
```

**Reply by DavidArno** [reply]
I find this thread fascinating, if a little too revealing of my own behaviour. Thread safety is - as HaloFour loves to say - clearly orthogonal to this feature request. But both he and Cyrus are obviously passionate about the the topic of thread safety and wanting to get more into the language. So when along comes a feature that could use it, they are all over it arguing that this feature should implement it. It's frustrating from my perspective as someone who's happy with the current "if you need thread safety, implement it" state of things to see the conversation hijacked in this way. But it's also a sobering thought that I do this with every feature that would benefit - in my view - from immutability and therefore others likely get equally frustrated with me for hijacking those conversations. Oh dear. 😳

**Reply by DavidArno** [reply]
@jnm2,

```cs
public List<string> Tags { get => field ??= new(); set; }
```
I find that a compelling use of `field`. I think that both satisfies this feature request and adds a good use-case to the field feature itself. Nice one.

**Reply by HaloFour** [reply]
> Thread safety is - as HaloFour loves to say - clearly orthogonal to this feature request.

I disagree, it's not orthogonal to this request because syntax that "simplifies" something like initialization is required to take a position on what kind of thread safety is employed when it comes to that initialization.  Even the decision to not be thread safe is a position on thread safety.  As noted, initialization within the language and runtime _is_ thread safe today.  Lazy initialization in the case of static locals in many languages, like VB.NET and C++ is also explicitly thread safe.  You can't tease apart thread safety and initialization.

Thread safety concerns aside, it's a weird way to go about lazy initialization.  Using compound assignment during an initializer doesn't make sense to me, anymore than `public int Count { get; set; } += 1;` would.  Why would something like that execute when calling the `get` accessor?  No initializer does that today.  What about `public int Count { get; set; } ??= 1;`?  What does that even mean?

This proposal seeks to enshrine one very opinionated approach to lazy initialization into the language, which will end up doing the wrong thing most of the time, due to the combination of being non-thread-safe, assuming values have sentinel values and not guarding against re-initialization if the value is manually set back to that sentinel.  It covers the very specific use case of OP, but that's not good enough for a language feature that is called "lazy initialization".

**Reply by CyrusNajmabadi** [reply]
>  It's frustrating from my perspective

Different people will have different opinions on what is important, and how things should work.  That's the purpose of that discussions, to share and attempt to understand all those views so we can (hopefully) design something as good as possible :-)

**Reply by popcatalin81** [reply]
> I disagree, it's not orthogonal to this request because syntax that "simplifies" something like initialization is required to take a position on what kind of thread safety is employed when it comes to that initialization

You're trying really really hard to conflate two very different things together which go by different rules, eager and lazy initialization, and insist they must obey the same rules when in fact in all programming languages and in all patterns they *do not*. This is your own bias, and your own way to see how this feature should be designed, simply because you've drawn a parallel starting from a naming coincidence: initialization. You could also replace initialization with "eager created objects" and "lazy created objects" and would be painfully obvious they would have different behaviors and expectations in regards to thread safety. 

> This proposal seeks to enshrine one very opinionated approach to lazy initialization

I don't think that something that is very common throughout the whole spectrum of programming languages is a very opinionated approach. Is the lazy initialization pattern thread safe? If you use the `??=` approach, no, if you use the `if ( foo== null) new ()`, then no, if you use a `Lazy<T>` then by default yes, but you can opt out. 

Is lazy initialization in C++ thread safe? No, you need to make it. Is it in C? No. Is lazy initialization in Go thread safe? No, you need to use the `sync.Once` mechanism to make it thread safe (opt-in). Is lazy initialization single execute in Haskell in a multithreaded program? No, it's single assignment but not single execute, in a race condition redundant work might be done, and results discarded, wildly different from single threaded initialization which is guaranteed single execute. 

There's not a single language that tries to force the same rules for eager vs lazy initialization considering multi threading behavior. Because for very practical reasons, they are not the same. 

> but that's not good enough for a language feature that is called "lazy initialization".

Only because you are trying to force some unreasonable expectations for two very different behaving concepts and try to force  them both into the same design space. But they are not, and should not be. Different expectations in regards to concurrency, different expectations  in regards to execution, different expectations in regards to memory utilization. Different uses cases in general, and different runtime behaviors. 

In fact the C# language is full of dual concepts, assumes different behaviors for related concepts: IE: classes - structs, copy by ref, copy by value, statics - instance, thread safe by default, not thread by default. The same is true for eager vs lazy.

**Reply by CyrusNajmabadi** [reply]
Personally, i would find that behavior surprising, and def a pitfall.  

>  and try to force them both into the same design space. But they are not, and should not be. 

That's a reasonable opinion to hold.  Though it's def going to differ depending on different people.  I personally think these are in the same design space, and would seek a solution without the problems brought about by keeping them separate.

**Reply by popcatalin81** [reply]
> That's a reasonable opinion to hold. Though it's def going to differ depending on different people. I personally think these are in the same design space, and would seek a solution without the problems brought about by keeping them separate.

Well, all I can say good luck! That's going to be one substantial challenge to undertake considering that AFAIK there are no examples out there of this being achievable even in concurrent languages.

**Reply by CyrusNajmabadi** [reply]
Honestly, if the threadsafety aspect is not a concern for you, I really think the `field ??= ...` approach would work fine.

**Reply by popcatalin81** [reply]
Yes `field ??= ` works perfectly fine for full properties currently because thread safety is not a concern (until it explicitly is for very particular cases)

**Reply by HaloFour** [reply]
> Is lazy initialization in C++ thread safe? No, you need to make it. Is it in C? No. Is lazy initialization in Go thread safe? No, you need to use the sync.Once mechanism to make it thread safe (opt-in). Is lazy initialization single execute in Haskell in a multithreaded program? No, it's single assignment but not single execute, in a race condition redundant work might be done, and results discarded, wildly different from single threaded initialization which is guaranteed single execute.
> 
> There's not a single language that tries to force the same rules for eager vs lazy initialization considering multi threading behavior. 

Of those languages that actually have a lazy initialization language feature, yes, they are generally thread safe.  Static locals in C and C++ are lazily initialized, and they are guaranteed to be thread safe.  Lazy initialization in Kotlin is guaranteed to be thread safe.  Lazy initialization in Scala is guaranteed to be thread safe.  Go does not have lazy initialization as a language feature, `sync.Once` is the Go equivalent of `Lazy<T>`.  Haskell is a different beast altogether.

Another important aspect of lazy initialization in all of those languages is that they do not rely on a sentinel value and they are guaranteed to execute exactly once.  It's not possible to flip the value back to "uninitialized" and trigger reinitialization.

**Reply by theunrepentantgeek** [reply]
@popcatalin81, lets assume for a moment that lazy initialization and concurrency really are actually orthogonal and independent concepts, as you maintain.

Feature design in C# is never done in isolation. 

One of the characteristics of C# is that changes are (for the most part) _well baked_. New features interact well with the rest of the language, with the BCL, and with the way the .Net ecosystem works. 

This happens because the language design folks take the time and care to consider how a proposed new feature interacts with existing (and, often with proposed) language features. 

There are many _seemingly good_ suggestions for C# that haven't proceeded because their interactions with other language features were found to be problemantic. This is an unqualified good.

Given this context, and with the way that concurrent programming is very much becoming the norm, it seems entirely reasonable to carefully consider the interaction between lazy-initialization and concurrency.

In fact, designing a lazy initialization feature _without_ due consideration of concurrency would be negligent.

**Comment by colejohnson66** [comment]
Does this have any advantage over `Lazy<T>` for the getter only cases?

```cs
using System.Threading; // for `LazyThreadSafetyMode`

private readonly Lazy<List<Category>> _categories = new(
    LazyThreadSafetyMode.None, // change to desired or remove to imply `None`
    () => new List<Category>() { Catgeory.Default });
public List<Category> Categories => _categories.Value;
```

**Reply by popcatalin81** [reply]
It's more lightweight than `Lazy<T>` with less allocation. `Lazy<T>` is a class which itself requires memory to allocate in addition to the fields values it's supposed to create and hold a reference to.

**Reply by theunrepentantgeek** [reply]
The overhead of an instance of `Lazy<T>` is very low. If that overhead exceeds the savings gained by making the initialization lazy, is there any point in making it lazy in the first place?

**Comment by svick** [comment]
Note that for collections that have empty default, you could make the property read-only and use collection initialization syntax:

```c#
public class Product
{
    public required string Name { get; set; }
    public List<string> Tags { get; } = new();
}

var p = new Product() 
{
    Name = "Peaches",
    Tags = { "Summer", "Sweet", "Fresh" }
};
```

**Reply by popcatalin81** [reply]
Yes, but I would still want the list to be lazy initialized ideally. 
Let's take and example, of a fairly typical domain object:
```csharp
public class Product
{
	public int Id { get; set; }
	public int Name { get; set; }
	public DateTime ExpirationDate { get; set; }
	// ....
	// omitted
	// ....
	// Reverse Navigation Properties
	public List<Category> Categories { get; set; } = new();
	public List<SubCategory> SubCategories { get; set; } = new();
	public List<MartketingCategory> MarketingCategories { get; set; } = new();
	public List<Location> Locations { get; set; } = new();
	public List<Product> RelatedProducts { get; set; } = new();
	public List<Tag> Tags { get; set; } = new();
	public List<Feature> Features { get; set; } = new();
	public List<PriceHistory> Prices { get; set; } = new(); 
	public List<Invetory> Invetories { get; set; } = new(); 
	public List<Store> Stores { get; set; } = new(); 
	public List<Photo> Photos { get; set; } = new();
	public List<Model> Models { get; set; } = new();
	public List<Document> Documents { get; set; } = new();
	public List<Employee> Supervisors { get; set; } = new();
	// ... etc
}
```
And you would use this domain object for some operation that loads a bunch of products (From DB or a WebService or some other place):
perform some changes, then save the changes:
```csharp
  using var unitOfWork = new UnitOfWork();
  foreach (var product in UnitOfWork.GetProducts(/* Some query */))
  {
  if (product.ExpirationDate < DateTime.Today.AddDays(7))
	  product.PromotionalPrice *= 0.5;
  }
  unitOfWork.SubmitChanges();
```

In this scenario, let's say you'd load 1000 products, for each product 14 empty new `List<T>()` would created eagerly and unnecessarily. That would make a single product instance allocate 840 bytes on creation. 14000 List objects and 728,000 byte allocated eagerly that re not used. 

However if each property becomes lazy initialized:

```csharp
    public List<Category> Categories 
    { 
        get => categories ??= new List<Category>(); 
        set => categories = value; 
    }
```
Then each product  would only take 112 bytes on creation, saving both memory and CPU cycles. 
This the main use case of lazy initialization, and that's why having the option for lazy initialization for lazy properties wold be a nice to have feature. In order to benefit from memory savings currently you have to refactor auto properties to use full properties and use the lazy initialization pattern, which is not ideal for simple domain objects, POCO clases or DTOs.

**Reply by CyrusNajmabadi** [reply]
> currently you have to refactor auto properties to use full properties and use the lazy initialization pattern, which is not ideal for simple domain objects, POCO clases or DTOs.

Or just autogenerate with an existing tool, or a simple SG.  WRite your fields, put some customization attributes on them (if you need customization. leave off if you don't), and just spit out all the common code in a different partial part.  Easy peasy (and fun!) :)

**Reply by popcatalin81** [reply]
> Or just autogenerate with an existing tool, or a simple SG. WRite your fields, put some customization attributes on them (if you need customization. leave off if you don't), and just spit out all the common code in a different partial part. Easy peasy (and fun!) :)

I did,
and changed hundreds of domain entities and thousands of properties to full properties with lazy initialization,  and got a docker process drop from ~12 GB peak to less than 2 GB, 20x less GC collections per hour and significant CPU utilization improvements. 

However now the previously nice domain model that was clear and easy to read, is bloated with code which is frankly visual noise.

**Reply by CyrusNajmabadi** [reply]
can you give an example of "is bloated with code which is frankly visual noise"?

I would expect you would simply write:

```c#
[DomainObject]
public class Product
{
	private int _Id;
	private int _Name;
	private DateTime _ExpirationDate;
```

And the rest would be taken care of for you.

**Reply by popcatalin81** [reply]
I understand that various options exist from code generators to  Roslyn Source Generators to AOP frameworks. But I'm not interested in that. I'm interested in a simple solution.
As solution where the choice between eager initialization or lazy initialization is done easily:

```csharp
public List<string> Tags { get; set; } = new(); // Eager 
public List<string> Tags { get; set; } ??= new(); // Lazy
```

**Comment by jnm2** [comment]
This was previously suggested in https://github.com/dotnet/csharplang/discussions/3247#discussioncomment-133554.
